distributed/service.py

```diff
@@ -0,0 +1,497 @@
+        key:
+            The key in the graph that created the `Service`. All instances created from
+            that key (across all workers) receive the same ``key``.
+
+            Multiple instances of the same type of `Service` can exist on a worker at
+            once, but all are guaranteed to have a different ``key``.
+        peers:
+            The IDs of all `Service` (including this one) currently being started to
+            handle this service task, and ServiceHandles to communicate with them.
+
+            Note that peer services are not guaranteed to be running yet.
+
+            The order of ``peers`` is undefined, but it will be the same for every
+            `Service` in ``peers``.
+        input_keys:
+            The input keys this service task will receive, *across all workers*, in
+            priority order.
+        output_keys:
+            The output keys this service task must produce, *across all workers*, in
+            priority order.
+        concierge:
+            A `Concierge` instance the `Service` can use to communicate with the
+            `Worker`. Used to hand off output keys back to Dask, restart or fail the
+            service task, etc.
+        leader:
+            Whether this instance was chosen to be the "leader" of the service task. The
+            scheduler will pick exactly one instance in ``peers`` to be the leader. The
+            order in which `Service` instances start up on workers is of course
+            undefined; the leader may not be the first instance to actually start.
+
+            The leader has no special treatment, and which instance is selected as the
+            leader is irrelevant to Dask. This flag is passed purely as a convenience
+            for Services, to make it easier to coordinate initialization tasks that must
+            run exactly once.
+        **kwargs:
+            Forward-compatibility only; any additional arguments can be ignored.
+        """
+        ...
+
+    # TODO: remove this in favor of `peer_joined` and `peer_left`.
+    # Once those methods are supported, and workers joining/leaving doesn't automatically
+    # restart the service task, we can't support `all_started` anymore: what if one of the
+    # original peers dies before it's started?
+    # `peer_joined` and `peer_left` are more powerful, because they allow for Service resizing,
+    # but still allow services to restart when workers leave (and wait for all to arrive) if
+    # they wish.
+    @abstractmethod
+    async def all_started(self) -> None:
+        """
+        Called once all instances in the original ``peers`` list have started.
+
+        Do not make the ``start`` method block until ``all_started`` has been called;
+        this will cause a deadlock.
+        """
+        ...
+
+    @abstractmethod
+    async def add_key(self, key: str, data: Any) -> None:
+        """
+        Called by the `Worker` to "hand off" data to the `Service`.
```

Comment:
I'm wondering if this "virtual tasks" approach may not be a good design, and we should do f1d977eb74e10e94b67377610deced11969d224f instead.
1. Allowing services to pump data into workers as fast as they want is basically root task overproduction. Services would have to come up with their own form of backpressure, waiting to add more keys until they ones they've already added have been used. But what if they're producing poor combinations of keys, and downstream tasks aren't getting what they need in order to run?

    Telling the scheduler "this is the menu of data available, you pick which ones you actually want" seems more correct in theory. While root-task overproduction is still a problem, it won't do any good (the scheduler just says "ALL OF THEM I LOVE KEYS!"), but in theory we could fix that in a couple different ways, and then the service model would fit nicely into it. Letting the scheduler pick which outputs are best to produce feels, in principle, like the right thing, even though right now the scheduler has no ability to make that decision.
2. It's just weird (and a little brittle) to have tasks in the graph that will never run?
3. It means we can't have triangle fusion (the outputs of two different service tasks can't be fused together). In practice, triangle fusion via Blockwise is an important protection against root task overproduction, because it allows DataFrame operations downstream of the `merge` to fuse into the `merge`. But triangle fusion is also hard to get right in the context of Services.

Just to beat a decayed and mummified horse over here, we wouldn't have to worry about triangle fusion if we fixed root task overproduction! We really have to twist ourselves into strange knots in many other places because of this scheduling error.